applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c

```diff
@@ -0,0 +1,821 @@
+	cJSON *acc_thres = NULL;
+
+	if (input == NULL) {
+		return -EINVAL;
+	}
+
+	root_obj = cJSON_Parse(input);
+	if (root_obj == NULL) {
+		return -ENOENT;
+	}
+
+	if (IS_ENABLED(CONFIG_CLOUD_CODEC_LOG_LEVEL_DBG)) {
+		json_print_obj("Decoded message:\n", root_obj);
+	}
+
+	group_obj = json_object_decode(root_obj, OBJECT_CONFIG);
+	if (group_obj != NULL) {
+		subgroup_obj = group_obj;
+		goto get_data;
+	}
+
+	group_obj = json_object_decode(root_obj, OBJECT_STATE);
+	if (group_obj == NULL) {
+		err = -ENODATA;
+		goto exit;
+	}
+
+	subgroup_obj = json_object_decode(group_obj, OBJECT_CONFIG);
+	if (subgroup_obj == NULL) {
+		err = -ENODATA;
+		goto exit;
+	}
+
+get_data:
+
+	gps_timeout = cJSON_GetObjectItem(subgroup_obj, CONFIG_GPS_TIMEOUT);
+	active = cJSON_GetObjectItem(subgroup_obj, CONFIG_DEVICE_MODE);
+	active_wait = cJSON_GetObjectItem(subgroup_obj, CONFIG_ACTIVE_TIMEOUT);
+	move_res = cJSON_GetObjectItem(subgroup_obj, CONFIG_MOVE_RES);
+	move_timeout = cJSON_GetObjectItem(subgroup_obj, CONFIG_MOVE_TIMEOUT);
+	acc_thres = cJSON_GetObjectItem(subgroup_obj, CONFIG_ACC_THRESHOLD);
+
+	if (gps_timeout != NULL) {
+		data->gps_timeout = gps_timeout->valueint;
+	}
+
+	if (active != NULL) {
+		data->active_mode = active->valueint;
+	}
+
+	if (active_wait != NULL) {
+		data->active_wait_timeout = active_wait->valueint;
+	}
+
+	if (move_res != NULL) {
+		data->movement_resolution = move_res->valueint;
+	}
+
+	if (move_timeout != NULL) {
+		data->movement_timeout = move_timeout->valueint;
+	}
+
+	if (acc_thres != NULL) {
+		data->accelerometer_threshold = acc_thres->valuedouble;
+	}
+
+exit:
+	cJSON_Delete(root_obj);
+	return err;
+}
+
+int cloud_codec_encode_config(struct cloud_codec_data *output,
+			      struct cloud_data_cfg *data)
+{
+	int err = 0;
+	char *buffer;
+
+	cJSON *root_obj = cJSON_CreateObject();
+	cJSON *state_obj = cJSON_CreateObject();
+	cJSON *rep_obj = cJSON_CreateObject();
+	cJSON *cfg_obj = cJSON_CreateObject();
+
+	if (root_obj == NULL || state_obj == NULL || rep_obj == NULL ||
+	    cfg_obj == NULL) {
+		cJSON_Delete(root_obj);
+		cJSON_Delete(state_obj);
+		cJSON_Delete(rep_obj);
+		cJSON_Delete(cfg_obj);
+		return -ENOMEM;
+	}
+
+	err += json_add_bool(cfg_obj, CONFIG_DEVICE_MODE,
+			     data->active_mode);
+	err += json_add_number(cfg_obj, CONFIG_GPS_TIMEOUT,
+			       data->gps_timeout);
+	err += json_add_number(cfg_obj, CONFIG_ACTIVE_TIMEOUT,
+			       data->active_wait_timeout);
+	err += json_add_number(cfg_obj, CONFIG_MOVE_RES,
+			       data->movement_resolution);
+	err += json_add_number(cfg_obj, CONFIG_MOVE_TIMEOUT,
+			       data->movement_timeout);
+	err += json_add_number(cfg_obj, CONFIG_ACC_THRESHOLD,
+			       data->accelerometer_threshold);
+
+	err += json_add_obj(rep_obj, OBJECT_CONFIG, cfg_obj);
+	err += json_add_obj(state_obj, OBJECT_REPORTED, rep_obj);
+	err += json_add_obj(root_obj, OBJECT_STATE, state_obj);
+
+	if (err) {
+		goto exit;
+	}
+
+	buffer = cJSON_PrintUnformatted(root_obj);
```

Comment:
Shouldn't it be checked if `buffer == NULL`? `cJSON_PrintUnformatted()` seems to use dynamic memory allocation, if it fails we'll end up with a NULL pointer, which may crash the app (in the `strlen()` call below for instance).

Same would apply in other places in this file where `cJSON_PrintUnformatted()` is called.

Reply:
Indeed, fixed buffer check and cleanup in case of failure for all occurences of `cJSON_PrintUnformatted()`